protocol_tests/connection/__init__.py

```diff
@@ -1,35 +1,87 @@
+"""Connection protocol messages and helpers."""
+
 import base64
+import json
 import re
 import uuid
+from collections import namedtuple
 
-from voluptuous import Schema, Optional
+from voluptuous import Schema, Optional, And, Extra, Match
 from aries_staticagent import Message, crypto
-from ..schema import MessageSchema
+from ..schema import MessageSchema, AtLeastOne
+
+
+TheirInfo = namedtuple(
+    'TheirInfo',
+    'endpoint, recipients, routing_keys'
+)
+
+
+class KeyReferenceError(Exception):
+    """Raised when no matching reference is found for a key."""
+
+
+class NoSuitableService(Exception):
+    """Raised when no DIDComm service is found."""
 
 
 class DIDDoc(dict):
     """DIDDoc class for creating and verifying DID Docs."""
+    # DIDDoc specification is very flexible: https://w3c-ccg.github.io
+    # This particular schema covers Ed25519 keys. All key types here:
+    # https://w3c-ccg.github.io/ld-cryptosuite-registry/
+    EXPECTED_SERVICE_TYPE = 'IndyAgent'
+    EXPECTED_SERVICE_SUFFIX = 'indy'
+
+    PUBLIC_KEY_VALIDATOR = Schema({
+        "id": str,
+        "type": "Ed25519VerificationKey2018",
+        "controller": str,
+        "publicKeyBase58": str
+    })
+
     VALIDATOR = Schema({
         "@context": "https://w3id.org/did/v1",
         "id": str,
-        "publicKey": [{
-            "id": str,
-            "type": "Ed25519VerificationKey2018",
-            "controller": str,
-            "publicKeyBase58": str
-        }],
-        "service": [{
-            "id": str,
-            "type": "IndyAgent",
-            "recipientKeys": [str],
-            Optional("routingKeys"): [str],
-            "serviceEndpoint": str,
-        }],
+        "publicKey": [PUBLIC_KEY_VALIDATOR],
+        # This is not fully correct; see:
+        # https://w3c.github.io/did-core/#authentication
+        Optional("authentication"): [
+            {
+                "type": "Ed25519SignatureAuthentication2018",
+                "publicKey": str
+            },
+            PUBLIC_KEY_VALIDATOR,
+            str
+        ],
+        "service": And(
+            # Service contains at least one agent service
+            AtLeastOne(
+                {
+                    'id': Match('.*{}$'.format(EXPECTED_SERVICE_SUFFIX)),
+                    'type': EXPECTED_SERVICE_TYPE,
```

Comment:
Does `id` really have to contain `indy` suffix?
Does `type` really have to match the constant?

DIDDod I found in this [HIPE](https://github.com/hyperledger/aries-rfcs/tree/master/concepts/0094-cross-domain-messaging#bobs-did-for-his-relationship-with-alice) doesn't pass validation

Reply:
I feel we need to enforce one or the other. Using `IndyAgent` would match many of the currently "interoperable" agents' approach whereas using `did-communication` would match the RFC you linked as well as the RFC on DIDDoc conventions. Now that you've called this out, I'm personally feeling like we should match what the RFCs say.